src/tests/profiler/native/handlesprofiler/handlesprofiler.cpp

```diff
@@ -0,0 +1,265 @@
+    {
+        hr = pCorProfilerInfo->CreateHandle(objectId, COR_PRF_HANDLE_TYPE::COR_PRF_HANDLE_TYPE_STRONG, &_strongHandle);
+        if (FAILED(hr))
+        {
+            _failures++;
+            printf("HandlesProfiler::ObjectAllocated: FAIL: CreateHandle failed for strong handle.\n");
+        }
+        else
+        {
+            printf("HandlesProfiler::ObjectAllocated: strong handle created.\n");
+        }
+    }
+    else
+    if (typeName == WCHAR("Profiler.Tests.TestClassForPinnedHandle"))
+    {
+        hr = pCorProfilerInfo->CreateHandle(objectId, COR_PRF_HANDLE_TYPE::COR_PRF_HANDLE_TYPE_PINNED, &_pinnedHandle);
+        if (FAILED(hr))
+        {
+            _failures++;
+            printf("HandlesProfiler::ObjectAllocated: FAIL: CreateHandle failed for pinned handle.\n");
+        }
+        else
+        {
+            printf("HandlesProfiler::ObjectAllocated: pinned handle created.\n");
+        }
+    }
+    return S_OK;
+}
+
+void HandlesProfiler::CheckIfAlive(ObjectHandleID handle, bool shouldBeAlive)
+{
+    if (handle == NULL)
+    {
+        _failures++;
+        printf("HandlesProfiler::CheckIfAlive: FAIL: null handle.\n");
+        return;
+    }
+
+    ObjectID objectId{0};
+    HRESULT hr = pCorProfilerInfo->GetObjectIDFromHandle(handle, &objectId);
+    if (FAILED(hr))
+    {
+        _failures++;
+        printf("HandlesProfiler::CheckIfAlive: FAIL: GetObjectIDFromHandle failed.\n");
+        return;
+    }
+
+    if (shouldBeAlive)
+    {
+        if (objectId == NULL)
+        {
+            _failures++;
+            printf("HandlesProfiler::CheckIfAlive: FAIL: the object should be alive.\n");
+        }
+        else
+        {
+            printf("HandlesProfiler::CheckIfAlive: object alive as expected ");
+            ClassID classId{0};
+            hr = pCorProfilerInfo->GetClassFromObject(objectId, &classId);
+            if (FAILED(hr))
+            {
+                _failures++;
+                printf("(FAIL: impossible to get class from object).\n");
+            }
+            else
+            {
+                String typeName = GetClassIDName(classId);
+                wcout << typeName.ToWString() << std::endl;
+            }
+        }
+    }
+    else
+    {
+        if (objectId != NULL)
+        {
+            _failures++;
+            printf("HandlesProfiler::CheckIfAlive: FAIL: the object should not be alive anymore.\n");
+        }
+        else
+        {
+            printf("HandlesProfiler::CheckIfAlive: object not alive as expected.\n");
+        }
+    }
+}
+
+HRESULT HandlesProfiler::GarbageCollectionFinished()
+{
+    SHUTDOWNGUARD();
+
+    if (!_isInduced)
+    {
+        return S_OK;
+    }
+
+    HRESULT hr = S_OK;
+    if (_gcCount == 1)
+    {
+        // weak should not be here anymore
+        CheckIfAlive(_weakHandle, false);
+
+        // the others should still be alive
+        CheckIfAlive(_strongHandle, true);
+        CheckIfAlive(_pinnedHandle, true);
```

Comment:
It would be good to add a check that the pinned handle never changes. Either checking the generation is stable or that the ObjectId returned never changes should suffice